source/common/common/assert.h

```diff
@@ -48,6 +48,28 @@ namespace Envoy {
   } while (false)
 #endif
 
+#ifndef NDEBUG
+#if __clang__
+#define __CLANG_PRAGMA(X) _Pragma(X)
+#else
+#define __CLANG_PRAGMA(X)
+#endif
+
+// clang-format off
+
+#define ASSERT_IGNORE_POTENTIALLY_EVALUATED(COND)                                                  \
+  do {                                                                                             \
+    __CLANG_PRAGMA("clang diagnostic push")                                                        \
+    __CLANG_PRAGMA("clang diagnostic ignored \"-Wpotentially-evaluated-expression\"")              \
```

Comment:
The comment in https://github.com/envoyproxy/envoy/pull/5243/files#diff-0bbc1b7ef8d5cfd41d41b2296caf7f1fR65 attempts to address the reason for this assert: the  `typeid(**mutable_config_providers_.begin())` statement triggers the potentially evaluated warning in clang.

This seems like a useful enough warning that I would rather not hide it by default (by folding it into the ASSERT() macro), but if others feel strongly about this I am certainly open to doing so.

Reply:
Hmm, if we do assert like this (evaluate the expression outside of `typeid`) then we don't need the pragma at all, since it is no longer potentially evaluated expression, do you think this is cleaner?

```C++
  ASSERT([&]() {
    if (!mutable_config_providers_.empty()) {
      const auto& first_provider = **mutable_config_providers_.begin();
      return typeid(*provider) == typeid(first_provider);
    }
    return true;
  }());
```

Reply:
This is a good solution; I definitely think it's cleaner overall even though it makes the ASSERT() slightly harder to parse. I just pushed this change and removed my changes to assert.h. Thanks!